File: src/y2025/d02/solution.clj
(ns y2025.d02.solution
  (:require [clojure.java.io :as io]
            [clojure.string :as s]))

(def sample-file "src/y2025/d02/sample-input.txt")
(def file "src/y2025/d02/input.txt")

(defn read-data
  [file]
  (with-open [rdr (io/reader file)]
    (into [] (line-seq rdr))))

(def range-separator ",")
(def id-separator "-")

(def sample-data
  (first (read-data sample-file)))

(def data
  (first (read-data file)))

;; Part 1
(defn invalid-id-p1?
  [id]
  (let [[first-half second-half] (split-at (/ (count id) 2) id)]
    (= first-half second-half)))

;; Part 2
(defn repeated-parts?
  [coll parts]
  (let [length (count coll)
        items-per-partition (quot length parts)
        partitions (partition-all items-per-partition coll)]
    (apply = partitions)))

(defn invalid-id-p2?
  [id]
  (let [start 2
        end (count id)]
    (->> (range start (+ end 1))                            ; all possible partitions
         (map #(repeated-parts? id %))                      ; which have repeating parts
         (some true?)                                       ; any of them have repeating parts?
         )))

;; both parts

(->> (s/split data (re-pattern range-separator))
     (map #(s/split % (re-pattern id-separator)))
     (map #(list (Long/parseLong (first %))
                 (Long/parseLong (last %))))
     (map (fn [[start end]] (cons end (range start end))))
     flatten
     (map str)
     ;(filter invalid-id-p1?)
     (filter invalid-id-p2?)
     (map #(Long/parseLong %))
     (apply +)
     )

File: nodejs/src/y2025/d02/solution.js
const fs = require('node:fs');
const ld = require('lodash');

const input = '/Users/tim/Documents/dev/advent-of-code/nodejs/src/y2025/d02/input.txt'
const sampleInput = '/Users/tim/Documents/dev/advent-of-code/nodejs/src/y2025/d02/sample-input.txt'

var data;

try {
  data = fs.readFileSync(input, 'utf8');
} catch (err) {
  console.error('Error reading the file: ' + err);
}

// Part 1

function toArray(str) {
    return str?.split("-")
}

function toRange([start, end]) {
    return ld.range(start, ++end);
}

function isEven(number) {
    return (number % 2) == 0;
}

function repeatedDigits(number) {
    var numStr = number.toString();
    var length = numStr.length;

    return isEven(length)
        && numStr.slice(0, length/2) == numStr.slice(length/2, length);
}

console.log(
    data.split(",")
        .map(toArray)
        .map(toRange)
        .flat(1)
        .filter(repeatedDigits)
        .reduce((accumulator, currentValue) => accumulator + currentValue)
);


// ------------------------

// Part 2

function splitStringInto(numStr, parts) {
    var length = numStr.length;
    var sizeOfEachPart = length / parts;

    var array = [];

    for(let i=0; i<length; i=i+sizeOfEachPart) {
        array = array.concat(numStr.slice(i,i+sizeOfEachPart));
    }

    return array;
}

function repeatedDigits2(number) {
    var numStr = number.toString();
    var length = numStr.length;

    if (length < 2) {
        return false;
    }

    var partitions = ld.range(2, length+1)
                       .filter((n) => length % n == 0);

    for (const part of partitions) {
        if (new Set(splitStringInto(numStr, part)).size == 1) {
            // Made up of repeating parts
            return true;
        }
    }

    // Not made up of repeating parts
    return false;
}

console.log(
    data.split(",")
        .map(toArray)
        .map(toRange)
        .flat(1)
        .filter(repeatedDigits2)
        .reduce((accumulator, currentValue) => accumulator + currentValue)
);


